frankenphp.c

```diff
@@ -652,106 +654,69 @@ static char *frankenphp_read_cookies(void) {
-  frankenphp_register_known_variable("GATEWAY_INTERFACE", known_variables[3],
-                                     track_vars_array);
-  frankenphp_register_known_variable("HTTP_HOST", known_variables[4],
-                                     track_vars_array);
-  frankenphp_register_known_variable("HTTPS", known_variables[5],
-                                     track_vars_array);
-  frankenphp_register_known_variable("PATH_INFO", known_variables[6],
-                                     track_vars_array);
-  frankenphp_register_known_variable("PHP_SELF", known_variables[7],
-                                     track_vars_array);
-  frankenphp_register_known_variable("REMOTE_ADDR", known_variables[8],
-                                     track_vars_array);
-  frankenphp_register_known_variable("REMOTE_HOST", known_variables[9],
-                                     track_vars_array);
-  frankenphp_register_known_variable("REMOTE_PORT", known_variables[10],
-                                     track_vars_array);
-  frankenphp_register_known_variable("REQUEST_SCHEME", known_variables[11],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SCRIPT_FILENAME", known_variables[12],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SCRIPT_NAME", known_variables[13],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SERVER_NAME", known_variables[14],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SERVER_PORT", known_variables[15],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SERVER_PROTOCOL", known_variables[16],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SERVER_SOFTWARE", known_variables[17],
-                                     track_vars_array);
-  frankenphp_register_known_variable("SSL_PROTOCOL", known_variables[18],
-                                     track_vars_array);
-
-  size_t new_val_len;
-  for (size_t i = 0; i < size; i++) {
-    if (sapi_module.input_filter(PARSE_SERVER, dynamic_variables[i].var,
-                                 &dynamic_variables[i].data,
-                                 dynamic_variables[i].data_len, &new_val_len)) {
-      php_register_variable_safe(dynamic_variables[i].var,
-                                 dynamic_variables[i].data, new_val_len,
-                                 track_vars_array);
-    }
+      request_uri, SG(request_info).request_uri, track_vars_array);
+}
+
+/* variables with user defined keys must be registered 'safely' */
+void frankenphp_register_variable_safe(char *key, char *val, size_t val_len,
+                                       zval *track_vars_array) {
+  size_t new_val_len = val_len;
+  if (!should_filter_var ||
+      sapi_module.input_filter(PARSE_SERVER, key, &val, new_val_len,
```

Comment:
I agree with you, I think we should go beyond PHP's security features. 

Other SAPIs mainly use this to enable `filter_input(PARSE_SERVER` (see [this nginxunit issue](https://github.com/nginx/unit/issues/854) as an example). Other SAPIs also don't care that much about its inefficiency since they don't necessarily register $_SERVER in each request (like is necessary in worker mode)
I don't think this can be optimized much without disabling the feature (via config) since it will duplicate all 'filtered' values.

Reply:
Maybe could we do something close to your initial proposal: call the filtering only one time (at bootime?) for known variables, and cache them?

Reply:
I found out that the globals filtering is actually [deprecated](https://php.watch/versions/8.1/filter.default-filter.default_options-deprecated). Using anything but `"filter.default: unsafe_raw"` in php.ini will emit a deprecation notice.
So calling the filter will do nothing but register it 3 times, once [here](https://github.com/php/php-src/blob/65f885738defe8c723e3d1131c0eb007cb71866d/ext/filter/filter.c#L332), once [here](https://github.com/php/php-src/blob/65f885738defe8c723e3d1131c0eb007cb71866d/ext/filter/filter.c#L348) and then again when we register it ourselves.

I verified this, tests will succeed even if we don't register variables ourselves as long as `sapi_module.input_filter(PARSE_SERVER` was called beforehand. I imagine other SAPIs just support this for backwards compatibility. It will probably be removed in an upcoming version (as using anything but `unsafe_raw ` was deprecated in 8.1 already and `unsafe_raw` does nothing to the actual value)

Reply:
Good news! So maybe could we skip this code path unless another value than `unsafe_raw` is passed?